tags/generators.md

```diff
@@ -0,0 +1,32 @@
+# Python Generators
+
+Often you'll need to return a list of values from a function.
+
+Imagine how you would implement a function that behaves like `map`,
+that is, it applies a function to each element in a list.:
+```python
+def my_map(function, old_list):
+    new_list = []
+    for element in old_list:
+        new_list.append(function(element))
+    return new_list
+
+for i in my_map(abs, [-1, -2, 3, -4]):
+    print(i)
+# 1, 2, 3, 4
+```
+However, there are two main issues with this approach:
+
+1. Initializing a list, appending to it and returning it is noisy boilerplate code.
+
+2. You will have to compute the result (which might be infinite) first, and this can take a lot of time and memory.
+
```

Comment:
I would mention the additional memory overhead associated with initializing and appending to the internal list; it's not just noisy boilerplate, there's also a significant cost associated with doing so.

It's also worth mentioning somewhere underneath this section when to use a list instead of a generator. Otherwise, a beginner might walk away from this assuming they should *always* use a generator, which is definitely not the case. If it needs to: be iterated over multiple times, accessed through specific indexes, or use any number of the methods that are available for lists (see https://docs.python.org/3.8/tutorial/datastructures.html#more-on-lists), a generator is not a suitable replacement.

Reply:
That's some good point, here's a quick benchmark, in my machine,

```py
import timeit
import os
import psutil

process = psutil.Process(os.getpid())
prev = 0

def current_memory(message: str):
    global prev
    current = process.memory_info().rss / 1024 / 1024
    print(message, f"| {current:,.2f} Mb ( +{current - prev:,.2f} )")
    prev = current

def get_gen():
    return (i for i in range(1_000_000))

def get_list():
    return [i for i in range(1_000_000)]

current_memory('Started.')

a = get_gen()
run_time = timeit.Timer(get_gen).timeit(10)
current_memory(f"Generator created, took {run_time:.3f}s")

b = get_list()
run_time = timeit.Timer(get_list).timeit(10)
current_memory(f"List created, took {run_time:.3f}s")
```
will print
```py
Started. | 14.66 Mb ( +14.66 )
Generator created, took 0.000s | 14.92 Mb ( +0.26 )
List created, took 0.654s | 54.46 Mb ( +39.54 )
```
I wonder if we should include small benchmarks somewhere to actually visualize the additional memory and performance overhead.
Also I agree with aeros, a better explanation on when and where to use / not use generator is very valuable. `generator` are super fast and efficient, but it does not mean you always have to use `generator`

Reply:
@aeros 
Isn't the time and memory overhead mentioned in the second item?
```
You will have to compute the result (which might be infinite) first, and this can take a lot of time and memory.
```
Maybe I can make it more explicit, though.

Reply:
Yeah, but there's a difference between the cost associated with getting all of results at once, vs the cost from initializing and appending to an internal list.

Reply:
@ikuyarihS 
```python
def get_gen():
    return (i for i in range(1_000_000))

def get_list():
    return [i for i in range(1_000_000)]
```
Between these two tests, I think there's some degree of bias towards the generator. Upon initialization, it doesn't have to compute any of the results (unlike the list, which it will do all at once). It might paint a more accurate picture if you compared the time to both initialize the comprehensions and iterate over the results. For example:

```python
def test_gen():
    total = 0
    gen = (i for i in range(1_000_000)
    for n in gen:
        total += n
    return total

def test_list():
    total = 0
    ls = [i for i in range(1_000_000)]
    for n in ls:
        total += n
    return total
```

Optionally, this could be simplified (and optimized for both) using the built-in `sum()` method:

```python
def test_gen():
    gen = (i for i in range(1_000_000))
    return sum(gen)

def test_list():
    ls = [i for i in range(1_000_000)]
    return sum(ls)
```

I suspect the former would be a better demonstration of the performance difference though, because the `sum()` method is efficient to the degree that it may overshadow some of the differences. But, the generator will be faster either way.

Good idea with the benchmark though, as long as it's simple enough for readers to understand (without being too much of a detraction), it would be a nice addition.

Reply:
I don't think that a benchmark would fit into a tag -- tags should be very short. I think this can be posted to the wiki (later).

Reply:
@decorator-factory 
> I don't think that a benchmark would fit into a tag -- tags should be very short. I think this can be posted to the wiki (later).

Good point. But we should at least mention the important performance differences between lists and generators, so that new users can walk away from this with a basic understanding of when to make use of either.